File: widgets/battery.js
const battery = await Service.import('battery')

const battery_dial = Widget.CircularProgress({
    className: 'battery-dial',
    rounded: false,
    inverted: false,
    startAt: 0.75,
    value: battery.bind('percent').as(p => p / 100),
    child: Widget.Label({
        className: "dial-icon",
        hexpand: true,
        setup: (self) => {
            self.hook(battery, (self) => {
                console.log(battery)
                const icons = [
                    ["󰂎", "󰁺", "󰁻", "󰁼", "󰁽", "󰁾", "󰁿", "󰂀", "󰂁", "󰂂", "󰁹"],
                    ["󰢟", "󰢜", "󰂆", "󰂇", "󰂈", "󰢝", "󰂉", "󰢞", "󰂊", "󰂋", "󰂅"],
                ];
                self.label = icons[Number(battery.charging)][Math.floor(battery.percent / 10)];
                self.tooltip_text = 'Battery ' + String(battery.percent) + '%';
            });
        }
    }),
    setup: (self) => {
        self.hook(battery, (self) => {
            if (battery.percent <= 30 && battery.charging === false) {
                self.toggleClassName("battery-low", true);
            } else {
                self.toggleClassName("battery-low", false);
            }
        });
    }
});

export {
    battery_dial
}

File: lua/widgets/bar/battery.lua
local Battery <const> = lgi.require("AstalBattery")

local battery <const> = Battery.get_default()

local icons <const> = {
  {"󰂎", "󰁺", "󰁻", "󰁼", "󰁽", "󰁾", "󰁿", "󰂀", "󰂁", "󰂂", "󰁹"},
  {"󰢟", "󰢜", "󰂆", "󰂇", "󰂈", "󰢝", "󰂉", "󰢞", "󰂊", "󰂋", "󰂅"},
};

return Widget.Box({
  class_name = 'battery-container',
  children = {
    Widget.CircularProgress({
      class_name = 'battery-dial',
      rounded = false,
      inverted = false,
      start_at = -.25,
      end_at = .75,
      value = bind(battery, 'percentage'),
      child = Widget.Label({
        halign = "CENTER",
        hexpand = true,
        justify = 2,
        setup = function(self)
          self:hook(battery, 'notify::percentage', function(self, percentage)
            self.label = icons[battery:get_charging() and 1 or 2][math.floor(percentage*10)]
          end)
          self.label = icons[battery:get_charging() and 2 or 1][math.floor(battery:get_percentage()*10)]
        end
      })
    })
  }
})
